### DISCUSSION TITLE
Use v8 snapshots

### DISCUSSION THREAD

**Discussion by piranna** [nodejs/single-executable#57] [Ideas]
Since we are creating a single executable application, source code is not intended to be modified, so it doesn't makes sense to just add a Javascript bundle to the Node.js executable. What not add instead a v8 snapshot? Code would be the same, but final file size will be smaller, start time shorter, and since we are going to target just the exact version of v8 included in the attached Node.js binary, there would not be runtime incompatibilities, in addition to make it more difficult to disassemble the original Javascript source code. There's similar previous aproaches with [bytenode](https://github.com/bytenode/bytenode) and [Hermes](https://hermesengine.dev/) engine.

**Comment by RaisinTen** [comment]
I think the reason why we opted for injecting the original JS file was because of simplicity but of course that can be changed since SEA is experimental.

### V8 snapshot

> What not add instead a v8 snapshot?

I'm assuming you're referring to https://nodejs.org/api/cli.html#--build-snapshot? The docs mention:
> Only a subset of the built-in modules work in the snapshot, though the Node.js core test suite checks that a few fairly complex applications can be snapshotted. 

Since a regular JS file supports all built-in modules and snapshots don't, it means that using the original JS file would have more capabilities for now.

---

Now let's look at the advantages this has over using a JS file:

> final file size will be smaller

I don't think that's always true. Take a look at this example of using a tiny JS file:
**index.js**
```js
require('v8').startupSnapshot.setDeserializeMainFunction(() => {
  const { stat } = require('fs/promises');

  (async () => {
    console.log(await stat('./index.js'));
  })()
});
```
If I create a snapshot file using `node --snapshot-blob snapshot.blob --build-snapshot index.js` and compare the sizes of the original JS file and the snapshot blob, this is what it looks like:
```sh
du -sh index.js snapshot.blob
4.0K	index.js
4.0M	snapshot.blob
```

I tried doing this on a real world js bundle, something we use at postman, which is 14 MB and building snapshots don't work for that because it relies on certain built-in modules that aren't supported in snapshots, like tty, child_process, vm, https, http, tls, readline, http2.

Also, it seems like building / using a snapshot fails when we use the `module` variable because it isn't defined in the context of a snapshot.

> start time shorter

That's true.

> make it more difficult to disassemble the original Javascript source code

If you inspect the snapshot blob in the example above, you'll be able to see the exact source code (you can search for `console.log(await`), so I don't see this as an advantage.

---

Usage - We can use this once we have a way to bake CLI options and a VFS (not implemented yet but there are plans to do that), so that we can bake something like `--snapshot-blob vfs://snapshot.blob` as a CLI option and read the blob from the VFS and execute it. Not sure if this can be loaded as a string though.

---

### V8 code cache

> There's similar previous aproaches with [bytenode](https://github.com/bytenode/bytenode) and [Hermes](https://hermesengine.dev/) engine.

I haven't dug into hermes yet but here's my assessment on the advantages of using the bytenode approach (which uses a code cache, not a snapshot, which is different) over a plan JS file:

> final file size will be smaller

For small files, like the one I showed above, the sizes are pretty much the same but it's slightly higher for the output file (specifically, 103 bytes for the input and 1008 bytes for the output).

Doesn't seem to be true for a large JS file either:
```sh
du -sh file.cjs file..jsc
 14M	index.js
 17M	index.jsc
```

So I wouldn't consider this to be an advantage.

> start time shorter

True.

> make it more difficult to disassemble the original Javascript source code

According to https://swarm.ptsecurity.com/how-we-bypassed-bytenode-and-decompiled-node-js-bytecode-in-ghidra, it seems like something that can be reverse engineered?

---

I can also think of a disadvantage:

I tried checking if exception info, like line number is preserved and it looks like that doesn't happen in bytenode, i.e., the stack trace of exceptions show that the source of the error is at `at evalmachine.<anonymous>:1:169` although the throw I have in my code is not anywhere close to line number 1. Not having that would probably make it significantly hard to debug applications. Are you aware of a way to solve this problem?

---

Usage - I think this can be used already if you inject an entry point that uses bytenode's API - [`bytenode.runBytecode(bytecodeBuffer)`](https://github.com/bytenode/bytenode#bytenoderunbytecodebytecodebuffer--any) to run a precompiled version of your application which you can specify in your source code as a string (could be base64 and decoded before passing to the API).

**Reply by joyeecheung** [reply]
> final file size will be smaller

This depends on the file specifically - the snapshot includes what has to be inside a Node.js heap + application states (note that it's serialized application state i.e. the result of running code, not code itself). So if the source file is smaller than the size of a basic Node.js heap, then it's going to be smaller (the snapshot of an empty script is 3.7MB i.e. the basic Node.js heap needs at least 3.7MB of space in the serialized blob. then what's on top are the application states). But generally source files can be bigger (because they are text). Take `test/fixtures/snapshot/typescript.js` for example, the file is 10MB, and the output of `node --build-snapshot test/fixtures/snapshot/typescript.js` is around 4.3MB, and that's because the size of source text of a TypeScript compiler is obviously non-trivial. I suspect this is the case for most real-world applications.

**Reply by joyeecheung** [reply]
Also this is the same regarding code cache - take the TypeScript bundle again for example,  the code cache is around 1.6MB (you can already check this out with `fs.writeFileSync('./cache.blob', new vm.Script(fs.readFileSync('test/fixtures/snapshot/typescript.js', 'utf8')).createCachedData())`.

**Reply by joyeecheung** [reply]
> Not having that would probably make it significantly hard to debug applications. Are you aware of a way to solve this problem?

This is the sacrifice that the application has to make for now if they want code cache **instead of** source text, not **in addition to** source text. V8 relies on reparsing the source to generate proper source positions. But it's also a common sacrifice that binaries make for smaller sizes (for example when you crash a release build of Chrome, it's very unlikely that you get to see any useful symbols in the stack trace, because they are mostly stripped away). Though I think it might be solvable by somehow plumbing additional source text to V8, like how PDB files work.

**Reply by piranna** [reply]
Option to run with a source map file, passed as parameter or environment variable?

**Comment by RaisinTen** [comment]
I wonder if @joyeecheung has more thoughts on this because she's leading the snapshots work?

Also, cc @nodejs/single-executable

**Reply by joyeecheung** [reply]
I don't see snapshot as replacements of JS entry points - they can be both supported. We could add something that tells the binary to either:

1. Start up from an injected JS file (the only things supported now)
2. Start up  from an embedded snapshot (that is produced by `node --build-snapshot` and then injected into the binary)
3. Start up from an embedded snapshot and then run an injected JS file (though I find this less useful considering you can also just put the code of the JS file into something that's run by the `v8.startupSnapshot.setDeserializeMainFunction()` callback in the snapshot script so that it gets baked into the snapshot too).

**Reply by joyeecheung** [reply]
I can give 2 a try (I think it should be fairly easy to implement). But I do have a question, where is this embedded snapshot supposed to be? Does it go into the `NODE_JS_CODE` (the name is kind of not suitable for this purpose, but oh well..) section? Or do we add a new section (and a new fuse)?

**Reply by joyeecheung** [reply]
POC here https://github.com/nodejs/node/pull/46824

**Reply by piranna** [reply]
Has there been any improvement with the PoC? On my use case, I have followed the path with bytenode (code caches instead of snapshots, sorry), and was able to create a tool to generate them fully for both executables and libraries, only missing part is to be able to include them in the Node.js binary (and the VFS for files that can't be bundled) to make them fully standalone :-)
